src/components/controllers/sensor_controller/sensor_controller.c

```diff
@@ -0,0 +1,220 @@
+// Copyright 2024 MACKI, Krzysztof Gliwinski
+
+#include "sensor_controller.h"
+
+#include <string.h>
+
+#include "adc_wrapper.h"
+#include "macki_log.h"
+#include "sensor_converters.h"
+#include "sensor_driver_definitions.h"
+
+#define TAG "SENSOR_CONTROLLER"
+
+typedef struct {
+  ads1115_driver_t* adc_expander;
+  lis2dw12_driver_t* accelerometer;
+  tmp1075_driver_t* temperature_sensor;
+  vl53l0x_driver_t* distance_sensor;
+} sensor_controller_drivers_t;
+
+static ring_buffer_t sensor_data_buffer;
+
+static sensor_controller_drivers_t sensor_controller_drivers = {
+    .adc_expander = &adc_expander,
+    .accelerometer = &accelerometer,
+    .temperature_sensor = &temperature_sensor,
+    .distance_sensor = &distance_sensor,
+};
+
+static sensor_controller_single_shot_data_t read_single_shot_data();
+static sensor_controller_continuous_data_t read_continuous_data();
+
+bool sensor_controller_init() {
+  rtc_wrapper_init();
+  init_shared_i2c_wrapper();
+
+  bool adc_ret = adc_wrapper_init();
+  if (!adc_ret) {
+    MACKI_LOG_ERROR(TAG, "Failed to initialize ADC wrapper");
+    return false;
+  }
+
+  ring_buffer_status_t rb_ret = ring_buffer_init(&sensor_data_buffer, 256);
```

Comment:
Huge buffer

Reply:
with a hidden feature of resetting one core 💪 💀